### PR TITLE
differentiate between unknown, true and false when propagate deployme…

### LINKED ISSUE #1332: Differentiate among "True", "False and "Unknown" when propagate Deployment Availability
**Problem**
Currently when we propagate deployment availability, we only set "true" or "false". https://github.com/google/knative-gcp/blob/87c676d2f0010ab6a6563829cef8c599a6254ef3/pkg/apis/intevents/v1beta1/pullsubscription_lifecycle.go#L88

**Exit Criteria**
Differentiate among "True", "False and "Unknown" when propagate Deployment Availability

### PR DESCRIPTION
**Pull request by capri-xiyue**
Fixes #1332 

## Proposed Changes
- 🎁 differentiate between unknown, true and false when propagate deployment status

### CODE REVIEW

File: pkg/apis/intevents/v1beta1/pullsubscription_lifecycle.go [hunk 1]
```diff
@@ -86,11 +86,20 @@ func (s *PullSubscriptionStatus) MarkNoSubscription(reason, messageFormat string
...
+				pullSubscriptionCondSet.Manage(s).MarkTrue(PullSubscriptionConditionDeployed)
+			} else if cond.Status == corev1.ConditionFalse {
+				pullSubscriptionCondSet.Manage(s).MarkFalse(PullSubscriptionConditionDeployed, cond.Reason, cond.Message)
+			} else if cond.Status == corev1.ConditionUnknown {
+				pullSubscriptionCondSet.Manage(s).MarkUnknown(PullSubscriptionConditionDeployed, cond.Reason, cond.Message)
+			}
+		}
+	}
+	if !deploymentAvailableFound {
 		pullSubscriptionCondSet.Manage(s).MarkFalse(PullSubscriptionConditionDeployed, "DeploymentUnavailable", "The Deployment '%s' is unavailable.", d.Name)
```
Comment (nachocano) [inline pkg/apis/intevents/v1beta1/pullsubscription_lifecycle.go:103]: should we mark it as unknown in this case?

Comment (capri-xiyue) [reply pkg/apis/intevents/v1beta1/pullsubscription_lifecycle.go:103]: Yep. We should mark it to unknown

**Comment by liu-cong** [comment]
The brokercell also uses similar logic for deployment status, which will be propagated to broker status.
https://github.com/google/knative-gcp/blob/4835f7b3a85e0d54e67d4633f8a4c3bfc069c343/pkg/apis/intevents/v1alpha1/brokercell_lifecycle.go#L94

Should we fix that as well?

File: pkg/apis/intevents/v1alpha1/brokercell_lifecycle.go [hunk 2]
```diff
@@ -91,10 +91,21 @@ func (bs *BrokerCellStatus) MarkIngressFailed(reason, format string, args ...int
...
-		brokerCellCondSet.Manage(bs).MarkTrue(BrokerCellConditionFanout)
-	} else {
-		brokerCellCondSet.Manage(bs).MarkFalse(BrokerCellConditionFanout, "DeploymentUnavailable", "Deployment %q is unavailable.", d.Name)
+	deploymentAvailableFound := false
+	for _, cond := range d.Status.Conditions {
+		if cond.Type == appsv1.DeploymentAvailable {
+			deploymentAvailableFound = true
+			if cond.Status == corev1.ConditionTrue {
+				brokerCellCondSet.Manage(bs).MarkTrue(BrokerCellConditionFanout)
+			} else if cond.Status == corev1.ConditionFalse {
```
Comment (liu-cong) [inline pkg/apis/intevents/v1alpha1/brokercell_lifecycle.go:100]: I just did a test, when you create a deployment, initially this is False until all pods are up and running for at least `minReadySeconds`. Therefore I think the behavior is still the same.

Note this is a general problem where clients give up when they see "False", not just for the deployment. I believe we could encounter this for other dependencies as well.

Maybe as a temporary solution we should just use True or Unknown.

Comment (capri-xiyue) [reply pkg/apis/intevents/v1alpha1/brokercell_lifecycle.go:100]: We should reflect the real status. If that's `False`, we should set status to `False`. CLI should implement the grace period.

File: pkg/apis/intevents/v1alpha1/brokercell_lifecycle.go [hunk 3]
```diff
@@ -91,10 +91,21 @@ func (bs *BrokerCellStatus) MarkIngressFailed(reason, format string, args ...int
 // to determine if BrokerCellConditionFanout should be marked as true or
 // false.
 func (bs *BrokerCellStatus) PropagateFanoutAvailability(d *appsv1.Deployment) {
-	if duck.DeploymentIsAvailable(&d.Status, false) {
-		brokerCellCondSet.Manage(bs).MarkTrue(BrokerCellConditionFanout)
-	} else {
-		brokerCellCondSet.Manage(bs).MarkFalse(BrokerCellConditionFanout, "DeploymentUnavailable", "Deployment %q is unavailable.", d.Name)
+	deploymentAvailableFound := false
```
Comment (liu-cong) [inline pkg/apis/intevents/v1alpha1/brokercell_lifecycle.go:94]: nit: Can you make this a helper function and avoid duplicating it many times?
```
func PropagateDeploymentStatus(d *appsv1.Deployment, manager ConditionManager, ct ConditionType)
```

And here you can just call `PropagateDeploymentStatus(d, brokerCellCondSet.manager(bs), BrokerCellConditionFanout `

File: pkg/apis/intevents/v1alpha1/test_helper.go [hunk 4]
```diff
@@ -74,6 +74,18 @@ func (t testHelper) UnavailableDeployment() *appsv1.Deployment {
 	return d
 }
 
+func (t testHelper) UnknownDeployment() *appsv1.Deployment {
+	d := &appsv1.Deployment{}
+	d.Name = "unknown"
+	d.Status.Conditions = []appsv1.DeploymentCondition{
+		{
+			Type:   appsv1.DeploymentAvailable,
+			Status: "Unknown",
```
Comment (liu-cong) [inline pkg/apis/intevents/v1alpha1/test_helper.go:83]: nit: use `corev1.ConditionUnknown`

File: pkg/apis/intevents/v1alpha1/brokercell_lifecycle.go (hunk 2 above)
Comment (liu-cong) [reply pkg/apis/intevents/v1alpha1/brokercell_lifecycle.go:100]: SG

**Pull request merged**